packages/keyring-controller/src/KeyringController.ts

```diff
@@ -790,7 +793,9 @@ export class KeyringController extends BaseController<
    * @returns A promise resolving to an array of addresses.
    */
   async getAccounts(): Promise<string[]> {
-    return this.#getAccountsFromKeyrings();
+    return this.#withControllerLock(async () =>
```

Comment:
Do we need a mutex here? This seems like a read-only method

Reply:
It is, but unlike the other read methods, it reads directly from the keyrings array instead of the controller state, so a consumer can potentially get an invalid array of addresses. I noticed this with some tests failing in https://github.com/MetaMask/core/pull/4192

Reply:
Perhaps we should discourage using this async method and promote `controller.state` instead

Reply:
Ah interesting. Good find, I would not have expected this to return invalid results mid-write operation, but that does make sense.

Maybe better to deprecate this instead of locking though. I could easily see someone thinking this was safe to use within a locking operation otherwise.

Reply:
Hmm OK. If you think locking this is safer, we can do that and deprecate it. That leaves us room to remove usages in a separate PR from the one that includes this update. I do think removing usage of this method is a good idea.

Reply:
Oh, refactoring this to return the accounts from state is an interesting idea as well, I like that.

Reply:
> Oh, refactoring this to return the accounts from state is an interesting idea as well, I like that.

I think I'll go with this!
 
> I could easily see someone thinking this was safe to use within a locking operation otherwise

Just to be sure on this, what is the specific concern here? a consumer calling this in an event listener, or another internal method that calls `this.getAccounts()`?

Reply:
I was thinking of internal callers with that comment. But refactoring this to use state solves that concern as well.